File: src/algorithm_java/leetcode/Problem581.java
package algorithm.leetcode;

public class Problem581 {

    public static int findUnsortedSubarray(int[] nums) {
        int begin = -1, end = -2;
        int max = nums[0], min = nums[nums.length - 1];
        for (int i = 1; i < nums.length; i++) {
            if (nums[i] > max) {
                max = nums[i];
            }
            if (nums[nums.length - i - 1] < min) {
                min = nums[nums.length - i - 1];
            }
            if (nums[i] < max)
                end = i;
            if (nums[nums.length - i - 1] > min)
                begin = nums.length - i - 1;
        }
        return end - begin + 1;
    }

    public static void main(String[] args) {
        int[] nums = new int[]{1,2,3,4};
        System.out.println(findUnsortedSubarray(nums));
    }

}

File: src/algorithm_cpp/problem581.cpp
class Solution {
public:
    int findUnsortedSubarray(vector<int>& nums) {
	int a = -1, b = -1;
	bool flag = false;
	for (int i = 1; i < nums.size(); i++) {
	    if (nums[i] < nums[i - 1]) {
		flag = true;
	    }
	    if (flag == true && (a == -1 || nums[i] < nums[a])) {
		a = i;
	    }	
	}
	flag = false;
	for (int i = nums.size() - 2; i >= 0; i--) {
	    if (nums[i] > nums[i + 1]) {
		flag = true;
	    }
	    if (flag == true && (b == -1 || nums[i] > nums[b])) {
		b = i;
	    }
	}
	if (a == -1 && b == -1) return 0;
	int c = 0, d = nums.size() - 1;
	for (int i = 0; i < nums.size(); i++) {
	    if (nums[i] > nums[a]) {
		c = i;
		break;
	    }
	}
	for (int i = nums.size() - 1; i >= 0; i--) {
	    if (nums[i] < nums[b]) {
		d = i;
		break;
	    }
	}
	return d - c + 1;
	
    }
};
